File: 297.py
class Codec:
    def __init__(self):
        self.cache = []
        self.i = 0;
        
    def serialize(self, root):
        if(not root):
            return "#"
        res = str(root.val)
        res += " " + self.serialize(root.left)
        res += " " + self.serialize(root.right)
        return res
        
    def des(self):
        val = self.cache[self.i]
        self.i += 1
        if(val == "#"):
            return None
        node = TreeNode(int(val))
        node.left = self.des();
        node.right = self.des();
        return node
    
    def deserialize(self, data):
        self.cache = data.split(" ");
        return self.des();

File: 297.cpp
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if(!root) return "*";
        std::string str = std::to_string(root -> val);
        str += (" " + serialize(root -> left));
        str += (" " + serialize(root -> right));
        return str;
    }

    // Decodes your encoded data to tree.
    TreeNode* des(std::stringstream& ss){
        std::string str;
        ss >> str;
        if(str == "*") return nullptr;
        TreeNode* node = new TreeNode(std::stoi(str));
        node -> left = des(ss);
        node -> right = des(ss);
        return node;
    }
    TreeNode* deserialize(string data) {
        std::stringstream ss;
        ss << data;
        return des(ss);
    }
};


